File: articles/aging-adaptation-and-development-in-later-life/python/model_aging_adaptation.py
#!/usr/bin/env python3
"""Model synthetic aging, adaptation, and development in later life."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "aging_adaptation_later_life_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_aging_adaptation_panel.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["id", "time"]).reset_index(drop=True)
    panel["lag_score"] = panel.groupby("id")["adjustment_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    adjustment_model = smf.ols(
        """
        adjustment_score ~ lag_score + time + functional_fit +
        current_support + current_adaptation + current_meaning +
        current_health + dignity_support + service_access
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    fit_model = smf.ols(
        """
        functional_fit ~ time + current_function + environmental_accessibility +
        current_function:environmental_accessibility + current_support +
        dignity_support + service_access
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_aging_adaptation_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC LATER-LIFE ADJUSTMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(adjustment_model.summary().as_text())
        f.write("\n\nFUNCTIONAL FIT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(fit_model.summary().as_text())

    panel["health_group"] = pd.qcut(
        panel["current_health"],
        3,
        labels=["Lower burden", "Moderate burden", "Higher burden"],
    )

    trajectory = panel.groupby(["time", "health_group"], as_index=False, observed=True).agg(
        average_adjustment=("adjustment_score", "mean"),
        average_functional_fit=("functional_fit", "mean"),
        standard_error=("adjustment_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["lower"] = trajectory["average_adjustment"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_adjustment"] + 1.96 * trajectory["standard_error"]
    trajectory.to_csv(OUTPUTS_DIR / "python_adjustment_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    for group_name, group in trajectory.groupby("health_group", observed=True):
        plt.plot(group["time"], group["average_adjustment"], marker="o", label=str(group_name))
    plt.xlabel("Time")
    plt.ylabel("Average adjustment score")
    plt.title("Synthetic Aging, Adaptation, and Later-Life Development")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_adjustment_trajectory.png", dpi=160)
    plt.close()

    fit_trajectory = panel.groupby("time", as_index=False).agg(
        average_functional_fit=("functional_fit", "mean"),
        average_function=("current_function", "mean"),
        average_support=("current_support", "mean"),
        average_health=("current_health", "mean"),
        average_adaptation=("current_adaptation", "mean"),
    )
    fit_trajectory.to_csv(OUTPUTS_DIR / "python_functional_fit_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(fit_trajectory["time"], fit_trajectory["average_functional_fit"], marker="o", label="functional fit")
    plt.plot(fit_trajectory["time"], fit_trajectory["average_support"], marker="o", label="support")
    plt.plot(fit_trajectory["time"], fit_trajectory["average_health"], marker="o", label="health burden")
    plt.xlabel("Time")
    plt.ylabel("Average index")
    plt.title("Synthetic Functional Fit, Support, and Health Burden")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_functional_fit_trajectory.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("adaptation_profile", as_index=False).agg(
        people=("id", "nunique"),
        average_adjustment=("adjustment_score", "mean"),
        average_functional_fit=("functional_fit", "mean"),
        average_support=("current_support", "mean"),
        average_health=("current_health", "mean"),
        average_adaptation=("current_adaptation", "mean"),
        average_meaning=("current_meaning", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_aging_adaptation_profiles.csv", index=False)

    print("Wrote Python aging adaptation model outputs.")


if __name__ == "__main__":
    main()

File: articles/aging-adaptation-and-development-in-later-life/r/model_aging_adaptation.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "aging_adaptation_later_life_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_aging_adaptation_panel.py first.")

panel <- read.csv(data_path)

adjustment_model <- lmer(
  adjustment_score ~ time + functional_fit + current_support +
    current_adaptation + current_meaning + current_health +
    dignity_support + service_access +
    (1 + time | care_context_id/id),
  data = panel
)

fit_model <- lmer(
  functional_fit ~ time + current_function + environmental_accessibility +
    current_function:environmental_accessibility + current_support +
    dignity_support + service_access +
    (1 + time | care_context_id/id),
  data = panel
)

capture.output(
  list(
    adjustment_model = summary(adjustment_model),
    functional_fit_model = summary(fit_model)
  ),
  file = file.path(outputs_dir, "r_aging_adaptation_model_summary.txt")
)

panel <- panel |>
  mutate(health_group = ntile(current_health, 3))

trajectory <- panel |>
  group_by(time, health_group) |>
  summarize(
    average_adjustment = mean(adjustment_score),
    average_functional_fit = mean(functional_fit),
    standard_error = sd(adjustment_score) / sqrt(n()),
    lower = average_adjustment - 1.96 * standard_error,
    upper = average_adjustment + 1.96 * standard_error,
    .groups = "drop"
  ) |>
  mutate(group = case_when(
    health_group == 1 ~ "Lower burden",
    health_group == 2 ~ "Moderate burden",
    TRUE ~ "Higher burden"
  ))

write.csv(trajectory, file.path(outputs_dir, "r_adjustment_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_adjustment_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_adjustment, linetype = group)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper, group = group), alpha = 0.12) +
    labs(title = "Synthetic Aging, Adaptation, and Later-Life Development", x = "Time", y = "Average adjustment", linetype = "Health group") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

fit <- panel |>
  group_by(time) |>
  summarize(
    average_functional_fit = mean(functional_fit),
    average_function = mean(current_function),
    average_support = mean(current_support),
    average_health = mean(current_health),
    average_adaptation = mean(current_adaptation),
    .groups = "drop"
  )

write.csv(fit, file.path(outputs_dir, "r_functional_fit_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_functional_fit_trajectory.png"),
  plot = ggplot(fit, aes(x = time)) +
    geom_line(aes(y = average_functional_fit, linetype = "functional fit"), linewidth = 1) +
    geom_line(aes(y = average_support, linetype = "support"), linewidth = 1) +
    geom_line(aes(y = average_health, linetype = "health burden"), linewidth = 1) +
    labs(title = "Synthetic Functional Fit, Support, and Health Burden", x = "Time", y = "Average index", linetype = "Measure") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R aging adaptation outputs.")
